tier4/Pilot.Auto
# ROS2 Porting: multi_object_tracker

# multi_object_tracker

## Summary

Convert the multi_object_tracker package to ROS2. A large but straightforward change with a few caveats which is why this PR is currently a draft.

Important (API) Changes:
* Convert `DataAssociation::assign` method return type to void due to `-Wreturn-type` error
* Pass `rclcpp::Time` by reference to `TrackerBase::getElapsedTimeFromLastUpdate` to perform duration calculation

Cosmetic Changes:
* Adding `#ifndef` guards on all header files and removing `#pragma once`
* Standardise use of `<>` vs `""` in `#include`
* Convert `.h` extension to `.hpp`
* Typos

## Testing

1. Checkout this branch and enter docker/ade environment

2. Build and source the repository
```
colcon build
. install/setup.bash
```
3. Launch file should launch without error
```
ros2 launch multi_object_tracker multi_object_tracker.launch.xml
```

## Discussion Points

### `lookupTransform`
The method `lookupTransform` in `tf2::Buffercore` no longer allows specification of lookup duration. At the moment I have left this implement to perform the lookup transform without a duration specification, see [issue here for further discussion](https://github.com/tier4/Pilot.Auto/issues/23).

**Update:** This has been implemented using the suggestion provided in the issue and taking inspiration from https://github.com/ros2/geometry2/blob/foxy/tf2_ros/test/test_buffer.cpp#L140-L158.

### `uuid_msgs` and `unique_id` Packages
This packages contains these dependencies however they have been ported (though the state of port is unclear) but not released on the ROS2 package index. I have commented on an issue tracking its progress regarding any further porting efforts.

**Update:** This has been implemented using the `unique_identifier_msgs` package which has a release available on foxy. I removed any references to Boost, using https://github.com/ros2/unique_identifier_msgs/blob/foxy/msg/UUID.msg instead. A different method of generate UUIDs using standard libraries was implemented instead of using the generate method in the `unique_id` package. Some references:
* https://github.com/ros2/rclcpp/blob/31c202e325476f6e912d6d23c17d13393156dbf4/rclcpp_action/src/client.cpp#L377-L379
* https://lowrey.me/guid-generation-in-c-11/
